File: src/ikc/main/Ioke.Lang/TypeCheckingNativeMethod.cs

namespace Ioke.Lang {
    using System.Collections;
    using System.Collections.Generic;

    using Ioke.Lang.Util;

    public class TypeCheckingNativeMethod : NativeMethod {
        public new class WithNoArguments : TypeCheckingNativeMethod {
            public WithNoArguments(string name, RawActivate activate) : base(name, TypeCheckingArgumentsDefinition.Empty(), activate) {}
            public WithNoArguments(string name, object mimic, RawActivate activate) : base(name, TypeCheckingArgumentsDefinition.EmptyButReceiverMustMimic(mimic), activate) {}
            public WithNoArguments(string name, object mimic, ArgsActivate activate) : base(name, TypeCheckingArgumentsDefinition.EmptyButReceiverMustMimic(mimic), activate) {}
        }
   
        public TypeCheckingNativeMethod(string name, TypeCheckingArgumentsDefinition arguments, ArgsActivate activate) : base(name, arguments, TypeCheckingRawActivate, activate) {}
        TypeCheckingNativeMethod(string name, TypeCheckingArgumentsDefinition arguments, RawActivate activate) : base(name, arguments, activate, null) {}

        private static object TypeCheckingRawActivate(IokeObject self, IokeObject context, IokeObject message, object on, NativeMethod outer) {
            IList args = new SaneArrayList();
            IDictionary<string, object> keywords = new SaneDictionary<string, object>();
            object receiver = ((TypeCheckingArgumentsDefinition)outer.ArgumentsDefinition).GetValidatedArgumentsAndReceiver(context, message, on, args, keywords);
            return outer.argsActivator(self, receiver, args, keywords, context, message);
        }
    }
}

File: src/ikj/main/ioke/lang/TypeCheckingNativeMethod.java
package ioke.lang;

import ioke.lang.exceptions.ControlFlow;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class TypeCheckingNativeMethod extends NativeMethod {
    public static class WithNoArguments extends TypeCheckingNativeMethod {
        private TypeCheckingArgumentsDefinition ARGUMENTS;

        public WithNoArguments(String name) {
            super(name);
            ARGUMENTS = TypeCheckingArgumentsDefinition.empty();
        }

        public WithNoArguments(String name, Object mimic) {
            super(name);
            ARGUMENTS = TypeCheckingArgumentsDefinition.emptyButReceiverMustMimic(mimic);
        }

        @Override
        public TypeCheckingArgumentsDefinition getArguments() {
            return ARGUMENTS;
        }
    }
   
    public TypeCheckingNativeMethod(String name) {
        super(name);
    }

    @Override
    public abstract TypeCheckingArgumentsDefinition getArguments();

    public Object activate(IokeObject self, Object on, List<Object> args,
            Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
        return super.activate(self, context, message, on);
    }

    @Override
    public Object activate(IokeObject self, IokeObject context,
            IokeObject message, Object on) throws ControlFlow {
        List<Object> args = new ArrayList<Object>();
        Map<String, Object> keywords = new HashMap<String, Object>();
        Object receiver = getArguments().getValidatedArgumentsAndReceiver(
                context, message, on, args, keywords);

        return activate(self, receiver, args, keywords, context, message);
    }
}
